File: backend/app/schemas/interview.py
"""Interview API schemas."""

import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class InterviewStartRequest(BaseModel):
    conversation_id: uuid.UUID | None = None
    position: str | None = None
    jd_text: str | None = None
    resume_text: str | None = None
    mode: Literal["full", "technical_interview"] = "full"


class InterviewAnswerRequest(BaseModel):
    message: str = Field(min_length=1)


class InterviewQuestionsRequest(BaseModel):
    position: str | None = None
    jd_text: str | None = None
    resume_text: str | None = None
    mode: Literal["full", "technical_interview"] = "full"


class InterviewSessionResponse(BaseModel):
    id: uuid.UUID
    conversation_id: uuid.UUID | None = None
    mode: str
    stage: str
    status: str
    position: str | None = None
    turns_in_stage: int = 0
    turn: dict[str, Any] | None = None
    review: dict[str, Any] | None = None
    evaluation: dict[str, Any] | None = None
    markdown: str
    created_at: datetime
    updated_at: datetime


class InterviewQuestionsResponse(BaseModel):
    questions: dict[str, Any]
    markdown: str

File: frontend/services/interview.ts
import { API_URL, apiHeaders, jsonHeaders } from "./api";
import type { InterviewSessionResponse, InterviewQuestionsResponse } from "@/types";

export async function startInterview(payload: {
  conversation_id?: string;
  position?: string;
  jd_text?: string;
  resume_text?: string;
  mode?: "full" | "technical_interview";
}): Promise<InterviewSessionResponse> {
  const response = await fetch(`${API_URL}/interview/start`, {
    method: "POST",
    headers: jsonHeaders(),
    body: JSON.stringify(payload),
  });
  if (!response.ok) {
    throw new Error((await response.text()) || `开始面试失败: ${response.statusText}`);
  }
  return response.json() as Promise<InterviewSessionResponse>;
}

export async function answerInterview(
  sessionId: string,
  message: string,
): Promise<InterviewSessionResponse> {
  const response = await fetch(`${API_URL}/interview/${sessionId}/answer`, {
    method: "POST",
    headers: jsonHeaders(),
    body: JSON.stringify({ message }),
  });
  if (!response.ok) {
    throw new Error((await response.text()) || `提交回答失败: ${response.statusText}`);
  }
  return response.json() as Promise<InterviewSessionResponse>;
}

export async function endInterview(sessionId: string): Promise<InterviewSessionResponse> {
  const response = await fetch(`${API_URL}/interview/${sessionId}/end`, {
    method: "POST",
    headers: apiHeaders(),
  });
  if (!response.ok) {
    throw new Error((await response.text()) || `结束面试失败: ${response.statusText}`);
  }
  return response.json() as Promise<InterviewSessionResponse>;
}

export async function getActiveInterview(
  conversationId?: string,
): Promise<InterviewSessionResponse | null> {
  const qs = conversationId ? `?conversation_id=${conversationId}` : "";
  const response = await fetch(`${API_URL}/interview/active${qs}`, {
    headers: apiHeaders(),
  });
  if (!response.ok) {
    throw new Error(`获取面试状态失败: ${response.statusText}`);
  }
  if (response.status === 204) return null;
  const data = await response.json();
  return data as InterviewSessionResponse | null;
}

export async function generateInterviewQuestions(payload: {
  position?: string;
  jd_text?: string;
  mode?: "full" | "technical_interview";
}): Promise<InterviewQuestionsResponse> {
  const response = await fetch(`${API_URL}/interview/questions`, {
    method: "POST",
    headers: jsonHeaders(),
    body: JSON.stringify(payload),
  });
  if (!response.ok) {
    throw new Error((await response.text()) || `生成题库失败: ${response.statusText}`);
  }
  return response.json() as Promise<InterviewQuestionsResponse>;
}
